### DISCUSSION TITLE
Livewire Component loosing data of child components

### DISCUSSION THREAD

**Discussion by alxhlz** [livewire/livewire#4724] [1. Help] [answered]
I am currently developing a new feature which should make it possible to manage firewall rules.
For that I created a component "RulesList" which holds an public property `$this->rules`. This collection contains all rules regarding the currently selected network and is getting loaded during `mount()`.

```php
public function mount()
{
        $this->rules = $this->network
            ->rules()
            ->orderBy('order')
            ->where('chain', $this->chain)
            ->where('comment', 'like', '%' . $this->search . '%')
            ->get();
}
``` 

In the blade part of the livewire component I am iterating through that collection and create a new nested livewire component for every item.

```php
@foreach($rules as $rule)
    @livewire('rule', ['rule' => $rule], key('ruleloop' . $loop->index))
@endforeach
```

Until here, everything is working just fine.
Now, I have added a button to the bottom of the "RulesList" to add completely new rules which technically is a `wire:click="addRule"` action that triggers the list component to create an empty rule and append it to the `$this->rules` collection.

```php
    public function addRule($chain = 'input')
    {
        if ($this->rules->count() > 0) {
            $newOrder = $this->rules->sortByDesc('order')->first()->order + 1;
        } else {
            $newOrder = 0;
        }

        $rule = new Rule;
        $rule->setConnection('mysql');
        $rule->network = $this->network;
        $rule->chain = $chain;
        $rule->action = 'accept';
        $rule->protocol = 'tcp';
        $rule->order = $newOrder;
        $this->rules->push($rule);
    }
```

This also works fine for the first rule that is going to be added by clicking the button. But as soon as I click the button again, the firstly created component fully looses it's data.

When I try to `dd()` at the beginning of the `addRule()` function, I am getting the following output.

```
2 => App\Models\Rule {[#786 ▼]()
      #connection: "mysql"
      #table: "rules"
      #primaryKey: "id"
      #keyType: "int"
      +incrementing: true
      #with: []
      #withCount: []
      +preventsLazyLoading: false
      #perPage: 15
      +exists: true
      +wasRecentlyCreated: false
      #escapeWhenCastingToString: false
      #attributes: array:17 [[▶]()]
      #original: array:17 [[▶]()]
      #changes: []
      #casts: []
      #classCastCache: []
      #attributeCastCache: []
      #dates: []
      #dateFormat: null
      #appends: []
      #dispatchesEvents: []
      #observables: []
      #relations: []
      #touches: []
      +timestamps: true
      #hidden: []
      #visible: []
      #fillable: array:1 [[▶]()]
      #guarded: array:1 [[▶]()]
    }
    3 => App\Models\Rule {[#669 ▼]()
      #connection: null
      #table: "rules"
      #primaryKey: "id"
      #keyType: "int"
      +incrementing: true
      #with: []
      #withCount: []
      +preventsLazyLoading: false
      #perPage: 15
      +exists: false
      +wasRecentlyCreated: false
      #escapeWhenCastingToString: false
      #attributes: []
      #original: []
      #changes: []
      #casts: []
      #classCastCache: []
      #attributeCastCache: []
      #dates: []
      #dateFormat: null
      #appends: []
      #dispatchesEvents: []
      #observables: []
      #relations: []
      #touches: []
      +timestamps: true
      #hidden: []
      #visible: []
      #fillable: array:1 [[▶]()]
      #guarded: array:1 [[▶]()]
    }
```

Here you can see, that already existing rules are completely fine but the newly created one is completely empty now.

Hopefully anyone has an idea that helps me fix that bug.

BR,
Alex

**Comment by joshhanley** [comment]
@alxhlz you will need to make sure you have white listed the properties of your model to make them available to the front end so they persist between requests. See the warning box in this section of the docs https://laravel-livewire.com/docs/2.x/properties#binding-models

Hope this helps!

**Reply by alxhlz** [reply]
I don't think that's the problem because in the child modules I already have the rules implemented and data that is coming directly from the database is working just fine.

The problem seems to be somewhere in the parents component `RulesList` that holds the list of `$this->rules`.

BR,
Alex

**Comment by joshhanley** [comment]
@alxhlz yeah but as those models already exist in the database, Livewire will pull the full data from there. Your new model isn't stored in the database though, so you need to tell Livewire which properties to keep track of for you using the rules property on the parent list component.

Also I wouldn't be using `$this->rules` as a property in your component, I would call it `$this->networkRules` or something, as `$rules` is already in use in Livewire as part of the validation system.

Hope this helps!

**Reply by alxhlz** [reply]
Good point, thanks! I renamed the variable to `firewallRules` as you suggested.

I also added a `$rules` property to the RulesList component (just to test) but that doesnt help. The problem still exists.

```php
    protected $rules = [
        'firewallRules' => 'nullable'
    ];
```

Is it the way to go to add new elements directly to the property that is initially filled from the database or would you suggest to do that differently?

**Reply by alxhlz** [reply]
I found out that I can add multiple new rules as long as there are no existing ones fetched from the database. Could the mixture be my problem?

**Reply by joshhanley** [reply]
@alxhlz your rules need to specify the individual properties, so like

```php
protected $rules = [
    'firewallRules.*.action' => 'nullable',
    'firewallRules.*.protocol' => 'nullable',
    ...
];
```

But yeah maybe you are right, it's possible having a collection with existing and new records will work properly.

**Comment by alxhlz** [comment, marked as answer]
I think I found a workaround for my problem. This includes the use of some behaviours that I can't fully explain by myself but for now I am rolling with it.

I seperated the firewallRules that I am fetching from the database and these ones I am newly creating. So in the RulesList view I have to call 2 foreach loops. For bootstrapping the rules I have to create empty models without any knowledge about it's database connection, because this kind of data wouldn't be persistent through multiple requests. I have to stick with data I have form fields and validation rules for.

This has the side effect, that I have to reload the page after saving newly created rules because atm I have no solution for moving newFirewallRules items to firewallItems collection. This would be necessary to receive the id of the model and the database connection to implement things like drag and drop for sorting them.

```php
@foreach($firewallRules as $rule)
    @livewire('virtfw.networks.rules.components.rule', ['rule' => $rule], key('ruleloop' . $loop->index . time()))
@endforeach
@foreach($newFirewallRules as $rule)
    @livewire('virtfw.networks.rules.components.rule', ['rule' => $rule, 'new' => true, 'index' => $loop->index], key('newruleloop' . $loop->index))
@endforeach
```

```php
    public function addRule($chain = 'input')
    {
        $rule = new VirtfwRule();
        $rule->network = $this->network;
        $rule->chain = $chain;
        $rule->action = 'accept';
        $rule->protocol = 'tcp';

        $this->newFirewallRules->push($rule);
    }
```

I hope it is some kind of clear what I did here. If I come up with a better solution I will keep you updated.
Thanks again @joshhanley for your input!

**Comment by joshhanley** [comment]
@alxhlz oh I didn't realise you were using nested components. So I personally wouldn't be using nested Livewire components for this. Instead I would just be using blade components for your rule output.

Especially if you also want to add drag and drop support to this, it will be much cleaner and easier to just track it all inside one Livewire component, which handles adding/removing and reordering.

But using models can also be fickle as you have found out.

Personally I would probably make use of arrays instead, so you can add, remove, reorder the data in your array as much as you need, then on save process the array and update any existing rules with their new order and create/delete any other rules as required.

Hope this helps!

**Reply by alxhlz** [reply]
Thank you again for that hint! I will consider to rebuild the component with the use of blade components instead of nested components. 🙇🏼

**Reply by joshhanley** [reply]
No worries 🙂
